Test anti-evil-maid

SergiiDmytruk opened the pull request:
Work in progress, but it somewhat works and there are things to be noted/clarified.

1. This is developed on an HDD image which already has AEM installed (or it can be installed by the test), but should be inserted somewhere in between other tests after installation later.
2. It uses QEMU with swtpm, but working with a hardware TPM will require a way to reset the TPM automatically.
3. `os-autoinst` might not start with a clean TPM state (newer versions do), so need to clean it after the test, but it doesn't work reliably (**EDIT2/4**: it doesn't work on test cancellation and on some failures).  ~Also might need to determine directory name for the case with multiple workers.~ (**EDIT4**: should be handled now.)
4. Current test assumes clean TPM state, so testing with and without SRK password seems to need two separate runs unless state reset can be done without restarting `swtpm` (~I think it didn't work for me at some point, but I didn't really try to make it work then~**EDIT4**: works for TPM2, but not for TPM1).
5. ~TOTP is in progress, I took some code from the Internet to not add dependencies to be installed which don't seem to be packaged by openSUSE.  Maybe the code is wrong, because it doesn't match at this point.~ **EDIT3**: code is fine, output parsing was wrong.  Used OCR to extract TOTP code on boot.
6. ~**EDIT**: I think that `anti-evil-maid` isn't installed by default, so will also need to do that.~ Not tested, but done.

SergiiDmytruk commented:
In current state this is capable of testing:
1. TPM 1.2 and TPM 2.0
2. Setup TPM with and without SRK password
3. MFA setup.

It's a single test file whose behaviour is controlled by `TEST_AEM_SRK_PASS` and `TEST_AEM_MFA` variables, so multiple runs are necessary to cover various cases and their combinations.

Issues:
1. At least with `swtpm` TPM1 doesn't seem to work after `tpm_clear` and reboot, so a single test can't verify setup with and without SRK password at least in case of QEMU.  TPM2 handles this better and can be reinitialized.
2. Hard-coding of `/dev/vda1` as boot device might be a problem, should it be determined?

@marmarek, any suggestions on integrating this properly into the test suite?  AEM affects boot process, so it probably should be run at the very end.  I see there are tests which use HDD, so multiple runs to check with/without SRK should be fine?

marmarek commented:
> AEM affects boot process, so it probably should be run at the very end

It doesn't really need to be at the end, it needs to be a separate job for AEM (and maybe some simple test after that as a sanity check). We have it split already (see [here for example](https://openqa.qubes-os.org/tests/79545#dependencies)), so it will be just few more jobs. Most start with a HDD of already installed system, but it seems it doesn't really change anything.

SergiiDmytruk commented:
I made a similar configuration with 9 new jobs.  In this case the first one was just to have a common parent and potentially install AEM once.  However, installation is probably better to do in `system_tests_update` which already prepares a common HDD image.

I'll post pieces here for discussion/example and also because importing `templates.json` into openQA didn't work (maybe wrong version).  So, under `scenarios` of relevant job templates:
```
    - aem
    - aem-tpm1
    - aem-tpm1-srk
    - aem-tpm1-mfa
    - aem-tpm1-srk-mfa
    - aem-tpm2
    - aem-tpm2-srk
    - aem-tpm2-mfa
    - aem-tpm2-srk-mfa
```
<details><summary>Test suites themselves (spoiler):</summary>
<p>


```
      {
         "description" : "Prepares base for all AEM tests",
         "name" : "aem",
         "settings" : [
            {
               "value" : "setup",
               "key" : "TEST_AEM"
            }
         ]
      },
      {
         "name" : "aem-tpm1",
         "settings" : [
            {
               "key" : "QEMUTPM",
               "value" : "instance"
            },
            {
               "key" : "QEMUTPM_VER",
               "value" : "1.2"
            },
            {
               "value" : "aem",
               "key" : "START_AFTER_TEST"
            },
            {
               "key" : "TEST_AEM",
               "value" : "1"
            }
         ]
      },
      {
         "settings" : [
            {
               "key" : "QEMUTPM",
               "value" : "instance"
            },
            {
               "key" : "QEMUTPM_VER",
               "value" : "1.2"
            },
            {
               "key" : "START_AFTER_TEST",
               "value" : "aem"
            },
            {
               "value" : "1",
               "key" : "TEST_AEM"
            },
            {
               "value" : "1",
               "key" : "TEST_AEM_SRK_PASS"
            }
         ],
         "name" : "aem-tpm1-srk"
      },
      {
         "name" : "aem-tpm1-mfa",
         "settings" : [
            {
               "value" : "instance",
               "key" : "QEMUTPM"
            },
            {
               "value" : "1.2",
               "key" : "QEMUTPM_VER"
            },
            {
               "key" : "START_AFTER_TEST",
               "value" : "aem"
            },
            {
               "key" : "TEST_AEM",
               "value" : "1"
            },
            {
               "value" : "1",
               "key" : "TEST_AEM_MFA"
            }
         ]
      },
      {
         "name" : "aem-tpm1-srk-mfa",
         "settings" : [
            {
               "value" : "instance",
               "key" : "QEMUTPM"
            },
            {
               "value" : "1.2",
               "key" : "QEMUTPM_VER"
            },
            {
               "value" : "aem",
               "key" : "START_AFTER_TEST"
            },
            {
               "value" : "1",
               "key" : "TEST_AEM"
            },
            {
               "key" : "TEST_AEM_MFA",
               "value" : "1"
            },
            {
               "key" : "TEST_AEM_SRK_PASS",
               "value" : "1"
            }
         ]
      },
      {
         "settings" : [
            {
               "value" : "instance",
               "key" : "QEMUTPM"
            },
            {
               "key" : "QEMUTPM_VER",
               "value" : "2.0"
            },
            {
               "key" : "START_AFTER_TEST",
               "value" : "aem"
            },
            {
               "key" : "TEST_AEM",
               "value" : "1"
            }
         ],
         "name" : "aem-tpm2"
      },
      {
         "name" : "aem-tpm2-mfa",
         "settings" : [
            {
               "value" : "instance",
               "key" : "QEMUTPM"
            },
            {
               "value" : "2.0",
               "key" : "QEMUTPM_VER"
            },
            {
               "value" : "aem",
               "key" : "START_AFTER_TEST"
            },
            {
               "value" : "1",
               "key" : "TEST_AEM"
            },
            {
               "value" : "1",
               "key" : "TEST_AEM_MFA"
            }
         ]
      },
      {
         "settings" : [
            {
               "value" : "instance",
               "key" : "QEMUTPM"
            },
            {
               "key" : "QEMUTPM_VER",
               "value" : "2.0"
            },
            {
               "key" : "START_AFTER_TEST",
               "value" : "aem"
            },
            {
               "value" : "1",
               "key" : "TEST_AEM"
            },
            {
               "key" : "TEST_AEM_SRK_PASS",
               "value" : "1"
            }
         ],
         "name" : "aem-tpm2-srk"
      },
      {
         "settings" : [
            {
               "key" : "QEMUTPM",
               "value" : "instance"
            },
            {
               "key" : "QEMUTPM_VER",
               "value" : "2.0"
            },
            {
               "key" : "START_AFTER_TEST",
               "value" : "aem"
            },
            {
               "value" : "1",
               "key" : "TEST_AEM"
            },
            {
               "key" : "TEST_AEM_MFA",
               "value" : "1"
            },
            {
               "value" : "1",
               "key" : "TEST_AEM_SRK_PASS"
            }
         ],
         "name" : "aem-tpm2-srk-mfa"
      }
```

</p>
</details> 

They are just possible combinations of TPM versions, SRK password and MFA.

Running them all a couple of times revealed several issues which were fixed except for a problem with MFA setup which happened only once.  Re-running tests multiple times didn't make it happen again, but I've added a `diag` so it's possible to see whether TOTP secret was extracted correctly from the terminal output (time should match, so maybe terminal output was partial or regexp picked up something wrong).

SergiiDmytruk commented:
The error I was getting isn't related to versions, job templates just need to be "cleared" and renamed for restoring to work properly.  No way seems to be provided for removing job templates, but dropping database does that :)

I've added:
* 8 test suites mentioned above
* `anti-evil-maid` installation to `update.pm`
* `aem-tpm1-srk-mfa` and `aem-tpm2-srk-mfa` to `Qubes OS pull requests` job group
* all 8 tests to `Qubes OS updates` job group

marmarek reviewed (changes requested):
After restructuring job templates it's a bit better, but I hit some more issues. See individual comments.

SergiiDmytruk commented on tests/aem.pm:
```diff
@@ -0,0 +1,242 @@
+        # accounting for possible delays by matching one of three areas
+        # against up to two codes
+        my $code = undef;
+        my @ocr = @{$needle->{ocr}};
+        for (my $i = 0; $i <= $#ocr; $i++) {
+            my $text = $ocr[$i];
+            # filtering after capturing more than needed to account for offsets
+            # caused by proportional font
+            $text =~ s/.*(\d{6}).*/$1/s;
+            if ($text eq $expected_old || $text eq $expected_new) {
+                $code = $text;
+                last;
+            }
+        }
```
`autoinst-log.txt` contains:
```
Can't exec "tesseract": Permission denied at /usr/lib/os-autoinst/ocr.pm line 20.
```
Which led to `$needle->{ocr}` not being initialized.  For some reason I didn't need to update AppArmor configuration to make OCR work, but I guess it might be necessary.

marmarek replied:
Indeed, default apparmor profile doesn't allow tesseract. Fixed.

SergiiDmytruk commented:
Updated according to comments and rebased.

marmarek commented on tests/aem.pm:
```diff
@@ -0,0 +1,242 @@
+}
+
+sub initiate_reboot {
+    type_string "reboot\n";
```
I meant after typing "reboot". In the test run, it returns and shell manages to print prompt one more time before actually rebooting.

SergiiDmytruk replied:
Corrected.  I thought it was typing something when shell wasn't yet expecting user input.

marmarek reviewed (approved):
Now it seems to work: https://openqa.qubes-os.org/tests/overview?distri=qubesos&version=4.2&build=2023082318-4.2&groupid=13
